In [ ]:
import pandas as pd
import numpy as np

### 1. What's the version of Pandas that you installed?

In [62]:
pd.__version__

'2.3.1'

### 2. How many records are in the dataset?

In [63]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [105]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0       152.0     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [104]:
df.shape

(9704, 11)

### Answer : 9704

### 3. How many fuel types are presented in the dataset?

In [65]:
df.fuel_type.nunique()

2

### Answer : 2

### 4. How many columns in the dataset have missing values?

In [66]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

### Answer : 4

### 5. What's the maximum fuel efficiency of cars from Asia?

In [67]:
df.groupby('origin').fuel_efficiency_mpg.max()

origin
Asia      23.759123
Europe    25.967222
USA       24.971452
Name: fuel_efficiency_mpg, dtype: float64

### Answer : 23.75

### 6. Median value of horsepower?
##### 1. Find the median value of horsepower column in the dataset.
##### 2. Next, calculate the most frequent value of the same horsepower column.
##### 3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
##### 4. Now, calculate the median value of horsepower once again.
##### 5. Has it changed?

In [68]:
df.horsepower.describe()

count    8996.000000
mean      149.657292
std        29.879555
min        37.000000
25%       130.000000
50%       149.000000
75%       170.000000
max       271.000000
Name: horsepower, dtype: float64

In [69]:
#Step 1
df.horsepower.median()

np.float64(149.0)

In [70]:
#Step 2
df.horsepower.mode()

0    152.0
Name: horsepower, dtype: float64

In [71]:
#Step 3
df['horsepower'] = df['horsepower'].fillna(152.0)

In [73]:
#Step 4
df.horsepower.median()

np.float64(152.0)


### Answer : Yes, it increased

### 7. Sum of weights
##### 1. Select all the cars from Asia
##### 2. Select only columns vehicle_weight and model_year
##### 3. Select the first 7 values
##### 4. Get the underlying NumPy array. Let's call it X.
##### 5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
##### 6. Invert XTX.
##### 7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
##### 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
##### 9. What's the sum of all the elements of the result?

In [117]:
#Step 1 : 
df_new = df[df['origin'] == 'Asia']

#Step 2, 3, 4 :
X = df_new[['vehicle_weight','model_year']].head(7).to_numpy()
print('X is:\n', X)

#Step 5 : 
XTX = X.T.dot(X)
print('\nXTX is:\n', XTX)

#Step 6 :
XTX_inv = np.linalg.inv(XTX)

#Step 7 : 
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

#Step 8 :
X1 = XTX_inv.dot(X.T)
w = X1.dot(y)
print('\n w is:\n', w)

#Step 9 :
result = w.sum()
print('\nThe sum of all elements of the result is:\n', result)

X is:
 [[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]

XTX is:
 [[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]

 w is:
 [0.01386421 0.5049067 ]

The sum of all elements of the result is:
 0.5187709081074016


### Answer : 0.51